Fancier teacher registration form

simoneduca opened the pull request.

Commit:
Add countries dropdown

Commit:
Add reusable checkbox component.

Add settings data.
TODO: enable multiple choice on checkbox.

eatyourgreens commented on src/presentational/Checkbox.jsx:
```diff
@@ -0,0 +1,47 @@
+  constructor(props) {
+    super(props);
+    this.state = {
+      selected: 'selected',
```
why is the initial value 'selected'? shouldn't it be `true` or `false`?

eatyourgreens commented on src/presentational/Checkbox.jsx:
```diff
@@ -0,0 +1,47 @@
+  render() {
+    const options = this.props.options.map((option) =>
+      <div>
+        <input
```
You should replace the `div` here with a `label` element, wrapping the input and its associated text label.

eatyourgreens commented on src/presentational/Checkbox.jsx:
```diff
@@ -0,0 +1,47 @@
+          value={option.label}
+          onChange={this.handleChange} />
+            {option.label}
+        <br />
```
Delete this `br` and use CSS to lay out your list of checkboxes.

eatyourgreens commented on src/presentational/Checkbox.jsx:
```diff
@@ -0,0 +1,47 @@
+      </div>
+    )
+    return (
+      <div>
```
I'd get rid of the nested `div` elements and mark this up as a `ul` list instead, with each input inside an `li`, then style the list with CSS to get either a vertical or inline layout.

eatyourgreens commented on src/presentational/Dropdown.jsx:
```diff
@@ -0,0 +1,46 @@
+  constructor(props) {
+    super(props);
+    this.state = {
+      selected: 'selected',
```
Initial state of selected should be `null` if nothing is selected?

eatyourgreens commented on src/presentational/Dropdown.jsx:
```diff
@@ -0,0 +1,46 @@
+        onChange={this.handleChange}>
+          {options}
+      </select>
+    )
```
Needs a `label` and the text label for the dropdown box eg. 'Country' before the `select`.

eatyourgreens commented on src/presentational/TeacherForm.jsx:
```diff
@@ -87,85 +93,82 @@ class TeacherForm extends Component {
-            value={this.state.country}
-            required
-            onChange={this.handleChange}/>
+            <label>Where do you teach?</label>
```
This label isn't associated with the dropdown. It might be better to make it part of the Dropdown component. (see above.)

eatyourgreens commented on src/presentational/TeacherForm.jsx:
```diff
@@ -87,85 +93,82 @@ class TeacherForm extends Component {
-            required
-            value={this.state.setting}
-            onChange={this.handleChange}/>
+            <label>In what educational setting do you plan to use this resource?</label>
```
See above. `label` elements need to be explicitly associated with a single form control. See the examples at https://developer.mozilla.org/en/docs/Web/HTML/Element/label

eatyourgreens commented on src/presentational/TeacherForm.jsx:
```diff
@@ -87,85 +93,82 @@ class TeacherForm extends Component {
-            value={this.state.setting}
-            onChange={this.handleChange}/>
+            <label>In what educational setting do you plan to use this resource?</label>
+            <Checkbox className="form-control"
```
Maybe call this component `CheckboxGroup` since it displays a list of checkboxes, not a single checkbox?

Also, each individual checkbox will require its own `label` but you can use `fieldset` to group several related controls, like these. https://developer.mozilla.org/en-US/docs/Web/HTML/Element/fieldset

eatyourgreens commented on src/presentational/TeacherForm.jsx:
```diff
@@ -87,85 +93,82 @@ class TeacherForm extends Component {
-            required
-            value={this.state.age}
-            onChange={this.handleChange}/>
+            <label>I plan to use this resource with my students.</label>
```
See above re. form labels

eatyourgreens commented on src/presentational/TeacherForm.jsx:
```diff
@@ -87,85 +93,82 @@ class TeacherForm extends Component {
-            required
-            value={this.state.course}
-            onChange={this.handleChange}/>
+            <label>In what course(s) do you plan to use this resource?</label>
```
Label needs to be associated with its form input.

eatyourgreens commented on src/presentational/TeacherForm.jsx:
```diff
@@ -87,85 +93,82 @@ class TeacherForm extends Component {
-            value={this.state.foundon}
-            onChange={this.handleChange}/>
+            <label>How did you find out about this resource?</label>
+            <input className="form-control"
```
Label needs to be associated with its form input.

eatyourgreens commented on src/presentational/TeacherForm.jsx:
```diff
@@ -87,85 +93,82 @@ class TeacherForm extends Component {
-            required
-            value={this.state.resources}
-            onChange={this.handleChange}/>
+            <label>Have you used HHMI BioInteractive resources other than WildCam Gorongosa in your teaching before?</label>
```
Label needs to be associated with its form input.

eatyourgreens commented on src/presentational/TeacherForm.jsx:
```diff
@@ -87,85 +93,82 @@ class TeacherForm extends Component {
-            required
-            value={this.state.feedback}
-            onChange={this.handleChange}/>
+            <label>Feedback from educators like you helps us improve our free educational resources. May we contact you at a later time?</label>
```
Label needs to be associated with its form input.

Commit:
Enable muliple choice checkbox.

Commit:
Add radio button group

Make all fields required on teacher form (html required doesn't work here)
Clear form properly
Spot of tidying up

Commit:
Fix form elements labels.

Commit:
Improve form accessibility and props naming.

Add react-a11y module.

Commit:
Resolve conflicts

simoneduca commented:
@eatyourgreens @shaunanoordin this is ready for review.

eatyourgreens commented on src/presentational/RadioButtonGroup.jsx:
```diff
@@ -0,0 +1,52 @@
+  handleChange(e) {
+    this.props.onChange(e);
+    this.setState({checked: e.target.checked});
+  }
```
Does `state.checked` ever get used? I can't see it anywhere else in this component.

eatyourgreens commented on src/presentational/CheckboxGroup.jsx:
```diff
@@ -0,0 +1,65 @@
+  handleChange(e) {
+    e.target.value = this.getCheckboxValues(e.target.name);
+    this.props.onChange(e);
+    this.setState({checked: e.target.checked});
```
Does `state.checked` ever get used? I can't see it anywhere else in this component.

eatyourgreens commented:
Looks good to me, apart from those comments about `state.checked` for the checkboxes and radio buttons.

Commit:
Remove unused code.

Refactor.
Add styling.

Commit:
Add emphasis.

simoneduca commented:
@eatyourgreens thanks! I removed `state.checked`, I forgot it there from the Dropdown component. This should be good to go.

eatyourgreens commented:
:+1:

Small style point: from the name, I'd expect `checkArray` to return `true` if the form was valid so I had to double-check that bit to make sure the code was checking the form properly.

simoneduca commented:
@eatyourgreens thanks! I could change it to `return array.some(item => item !== null && item !== '')` if it makes it clearer?

eatyourgreens commented:
@simoneduca `return !array.some(item => item === null || item === '')` would flip the true/false value of the return value, but it's a minor point.

Your updated check will return `true` if some fields are filled out, but not all. I don't think that's what you want.
